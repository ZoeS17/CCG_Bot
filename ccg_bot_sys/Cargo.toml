[package]
authors = ["Zoe Kahala <>"]
name = "ccg_bot_sys"
license = "EUPL-1.2"
repository = "https://github.com/ZoeS17/CCG_Bot.git"
version = "0.1.0"
edition = "2021"
publish = false

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
proc-macro2 = "1.0.51"
quote = "1.0.23"
syn = "1.0.109"

[lib]
proc-macro = true