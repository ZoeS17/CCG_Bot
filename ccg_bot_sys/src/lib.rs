use proc_macro::TokenStream;
use syn::{parse_macro_input, Item};

mod unstable;

#[proc_macro_attribute]
pub fn unstable(args: TokenStream, item: TokenStream) -> TokenStream {
    let args = parse_macro_input!(args as syn::AttributeArgs);
    let attr = unstable::UnstableAttribute::from(args);

    match parse_macro_input!(item as Item) {
        Item::Type(item_type) => attr.expand(item_type),
        Item::Enum(item_enum) => attr.expand(item_enum),
        Item::Struct(item_struct) => attr.expand(item_struct),
        Item::Fn(item_fn) => attr.expand(item_fn),
        Item::Mod(item_mod) => attr.expand(item_mod),
        Item::Trait(item_trait) => attr.expand(item_trait),
        Item::Const(item_const) => attr.expand(item_const),
        Item::Static(item_static) => attr.expand(item_static),
        _ => panic!("unsupported item type"),
    }
}
