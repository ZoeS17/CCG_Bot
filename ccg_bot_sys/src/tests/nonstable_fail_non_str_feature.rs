use ccg_bot_sys::nonstable;

#[nonstable(feature = 1)]
static NONSTABLE_STATIC_WITH_NON_STR_FEATURE: i32 = 0_i32;

pub fn main() {
    let _ = NONSTABLE_STATIC_WITH_NON_STR_FEATURE;
}