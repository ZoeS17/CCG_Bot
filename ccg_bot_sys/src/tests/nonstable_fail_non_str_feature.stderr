error: custom attribute panicked
 --> src/tests/nonstable_fail_non_str_feature.rs:3:1
  |
3 | #[nonstable(feature = 1)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: explicit panic

error[E0425]: cannot find value `NONSTABLE_STATIC_WITH_NON_STR_FEATURE` in this scope
 --> src/tests/nonstable_fail_non_str_feature.rs:7:13
  |
7 |     let _ = NONSTABLE_STATIC_WITH_NON_STR_FEATURE;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
